diopter/sanitizer.py

```diff
@@ -363,6 +372,67 @@ def check_for_ub_and_address_sanitizer_errors(
                 print("Sanitizer checks passed")
             return SanitizationResult()
 
+    def check_for_memory_sanitizer_errors(
+        self, program: SourceProgram
+    ) -> SanitizationResult:
+        """Checks the program for memory sanitizer errors.
+
+        Compiles program with self.clang and -fsanitize=memory, then
+        it runs the checked program and reports whether it failed (or timed out).
+
+        Args:
+            program (SourceProgram):
+                The program to check.
+        Returns:
+            SanitizationResult:
+                whether the program failed sanitization or not.
+        """
+
+        with TempDirEnv():
+            # Compile program with -fsanitize=...
+            try:
+                result = CompilationSetting(
+                    compiler=self.clang,
+                    opt_level=self.sanitizer_opt_level,
+                ).compile_program(
+                    program,
+                    ExeCompilationOutput(None),
+                    (
+                        "-Wall",
+                        "-Wextra",
+                        "-Wpedantic",
+                        "-Wno-builtin-declaration-mismatch",
+                        "-fsanitize=memory",
+                        "-fno-sanitize-recover=all",
+                        "-O0",
```

Comment:
The optimization level is set in `CompilationSetting` (a few lines above)

Reply:
Yes, this is indended as clang at -O1 and above will aggressively optimize away the ub for the given test case

Reply:
It's really nice to have a parameter that controls the optimization level and then hardcode another one. Would it make sense to make `O0` the default for all sanitizers?

Reply:
Instead of passing `-O0` here, it makes more sense to change the default value of `sanitizer_opt_level: OptLevel = OptLevel.O0` in `__init__`

Reply:
resolved